### PR TITLE
Allow control over the protocol version

### PR DESCRIPTION
**Pull request by mgravell**
The *aim* here is to allow us to reliably talk to HTTP/2 servers, but: there are a few extra bits involved too.

Primary API change:

- new `WithProtocolVersion` extension method that simply sets the `.Message.Version`

However, that isn't enough to work reliably / validate this, so a range of secondary changes are included:

- add `netcoreapp3.1` TFM to the test project that uses Kestrel to spin up a server with all the scenarios (HTTP1.1, HTTP2, HTTP1.1+HTTP2, all with/without TLS)
- add a self-signed test certificate to the test project (with notes on how to update as needed)
- add `.ServerCertificateCustomValidationCallback` option to `HttpSettings` for cert validation (so we can reliably test against the self-signed certificate)
- add `HttpSettings.GlobalAllowUnencryptedHttp2` to toggle the global "http2 without TLS" setting
- add 24 test expectations about what happens when you use HTTP/2

- (unrelated) not all cancellations are `TaskCanceledException`; prefer `OperationCanceledException` instead

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -1,8 +1,11 @@
-image: Visual Studio 2017
+image: Visual Studio 2019
 
 init:
   - git config --global core.autocrlf input
 
+install:
+  - choco install dotnetcore-sdk --version 3.0.100
```
Comment (NickCraver) [inline appveyor.yml]: Can remove this - project is using the (newer) one in the image anyway, so this is just eating time on the build really

File: tests/Directory.Build.props [hunk 2]
```diff
@@ -7,7 +7,7 @@
     <DebugType>Full</DebugType>
     <xUnitVersion>2.4.1</xUnitVersion>
   </PropertyGroup>
-  
+
```
Comment (NickCraver) [inline tests/Directory.Build.props]: Can revert: `git checkout master -- tests/Directory.Build.props`

File: src/StackExchange.Utils.Http/Http.cs [hunk 3]
```diff
@@ -108,5 +108,15 @@ internal static async Task<HttpCallResponse<T>> SendAsync<T>(IRequestBuilder<T>
...
+
+        private const string Switch_AllowUnencryptedHttp2 = "System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport";
+        /// <summary>
+        /// Allows HttpClient to use HTTP/2 without TLS
+        /// </summary>
+        public static bool AllowUnencryptedHttp2
+        {
+            get => AppContext.TryGetSwitch(Switch_AllowUnencryptedHttp2, out var enabled) && enabled;
+            set => AppContext.SetSwitch(Switch_AllowUnencryptedHttp2, value);
+        }
```
Comment (NickCraver) [inline src/StackExchange.Utils.Http/Http.cs]: Definitely not a fan of the static, at least on this class (it's on the default intellisense for every intended usage, etc.) Perhaps on `HttpSettings` and `GloballyAllow...`?

Comment (mgravell) [reply src/StackExchange.Utils.Http/Http.cs]: It is fundamentally a global thing; static on `HttpSettings` work for you?

File: src/StackExchange.Utils.Http/HttpSettings.cs [hunk 4]
```diff
@@ -78,5 +85,15 @@ public HttpSettings()
             // Default pool by default
             ClientPool = new DefaultHttpClientPool(this);
         }
+
+        private const string Switch_AllowUnencryptedHttp2 = "System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport";
+        /// <summary>
+        /// Allows HttpClient to use HTTP/2 without TLS
```
Comment (NickCraver) [inline src/StackExchange.Utils.Http/HttpSettings.cs]: Let's note this is *app global* and affects more than this library? I'd prefer that be in both the description and property name to be super clear - remember how confusing `ServicePointManager` is to people...

**Review by NickCraver — APPROVED** [review]
:+1:

**Pull request merged**